File: python/leetcode59.py
class Solution:
    def generateMatrix(self, n: int) -> [[int]]:
        (
            l,
            r,
            t,
            b,
        ) = (
            0,
            n - 1,
            0,
            n - 1,
        )
        mar = [[0 for _ in range(n)] for _ in range(n)]
        num, tar = 1, n * n
        while num <= tar:
            for i in range(l, r + 1):
                mar[t][i] = num
                num += 1

            t += 1
            for i in range(t, b + 1):
                mar[i][r] = num
                num += 1
            r -= 1
            for i in range(r, l - 1, -1):  # right to left
                mar[b][i] = num
                num += 1
            b -= 1
            for i in range(b, t - 1, -1):  # bottom to top
                mar[i][l] = num
                num += 1
            l += 1
        return mar

File: C++/leetcode59.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution
{
public:
    vector<vector<int>> generateMatrix(int n)
    {
        vector<vector<int>> matrix(n);
        for (int i = 0; i < matrix.size(); i++)
            matrix[i].resize(n);
        // 上下左右
        int u = 0;
        int d = n - 1;
        int l = 0;
        int r = n - 1;
        int num = 1;

        while (true)
        {
            // 上
            for (int i = l; i <= r; i++)
                matrix[u][i] = num++;
            if (u++ >= d)
                break;
            // 右
            for (int i = u; i <= d; i++)
                matrix[i][r] = num++;
            if (r-- <= l)
                break;
            // 下
            for (int i = r; i >= l; i--)
                matrix[d][i] = num++;
            if (d-- <= u)
                break;
            // 左
            for (int i = d; i >= u; i--)
                matrix[i][l] = num++;
            if (l++ >= r)
                break;
        }
        return matrix;
    }
};

int main()
{
    Solution solution;
    int n = 3;
    vector<vector<int>> ans = solution.generateMatrix(n);
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}